### ISSUE TITLE
[Enhancement]: Asynchronous Refund API

### ISSUE THREAD

**Issue by Biont** [woothemes/woocommerce#52338]
### Describe the solution you'd like

The current APIs surrounding refunds assume that refunds from external payment service providers can be carried out synchronously:

`WC_Payment_Gateway::process_refund` returns `bool|WP_Error`, meaning a refund can only fail or succeed. You can customize an error message by returning a WP_Error object in case of a failed refund. 
The process does not leave room for a refund to stay in a pending state for an indefinite amount of time.

Modern PSP often offer a wide range of payment methods with vastly differing paradigms and functionalities (wallets, credit cards, bank transfer) and there is an array of systems involved with each of them that need to be communicated with. Granted: This is all transparent to the consumer of the PSP's API but still, creating a synchronous refund API is likely a monumental if not impossible task.

Therefore the existing workflows for WooCommerce refunds are hard to integrate into the workflows of PSP APIs. 
Please consider adding an asynchronous route for refunds.

### Describe alternatives you've considered

The current workaround is to abuse the WP_error code path to prevent WooCommerce from finalizing the refund. An error message like *"Your refund request has been sent and is now being processed by the PSP"* can be returned so that the refund is intercepted by WC without showing a distracting error to the merchant.

[`WC_Payment_Gateway::process_refund` also does not receive the WP_Refund object](https://github.com/woocommerce/woocommerce/blob/1fde6a54de95eb7d778e881343391f0241ae0d6b/plugins/woocommerce/includes/abstracts/abstract-wc-payment-gateway.php#L440-L442), so in order to apply the refund at a later point, developers are currently forced to look for other ways:

By hooking into `woocommerce_create_refund`, the refund object can be obtained and stored in order to carry out the refund as soon as the PSP signals success. This is a complex setup and making this functionality more streamlined and approachable would make the interaction with modern payment services more stable and convenient.

Summing up, the process is basically to intercept and abort the synchronous refund, serialize + store it. Then send off the API request to the PSP. When a webhook about the successful refund is received, deserialize the refund data and programmatically apply it.

### Additional context

Making the `WC_Refund` object accessible to the payment gateway appears to be a pivotal component of this proposal. Please let me know if you want me to create a separate issue for this in case you decide to pick up this issue.
